### ISSUE TITLE
Compute delp and hydrostatic delz increments in jediinc2fv3.py

### ISSUE THREAD

**Issue by RussTreadon-NOAA** [NOAA-EMC/GDASApp#135]
This issue is opened to document the addition of `delp` and hydrostatic `delz` increment calculations to `ush/jediinc2fv3.py`.   Below is background information motivating creation of this issue.

The operational GFS GSI/EnKF pass a hydrostatic `delz` increment to the FV3GFS forecast.  For example, the EnKF code for this calculation is shown in [this comment](https://github.com/JCSDA-internal/fv3-jedi/issues/602#issuecomment-1241139277) from fv3-jedi issue [#602](https://github.com/JCSDA-internal/fv3-jedi/issues/602).

Code was added to `fv3-jedi/src/fv3jedi/State/fv3jedi_state_mod.F90` to mimic this calculation.   PR [#514](https://github.com/JCSDA-internal/fv3-jedi/pull/514) refactored this source code file.   For example, the following `delp` section of code is no longer present:
```
! Convert ps to delp if necessary
! -------------------------------
if (hasfield(rhs_fields, 'ps')) then !ps in increment
  if (.not.hasfield(rhs_fields, 'delp')) then !delp not in increment
    if (self%has_field('delp')) then !delp in state
      allocate(rhs_delp(isc:iec,jsc:jec,1:npz))
      call get_field(rhs_fields, 'ps', rhs_ps)
      do k = 1, npz
        rhs_delp(:,:,k) = (geom%bk(k+1)-geom%bk(k))*rhs_ps(:,:,1) !TLM
      enddo
    endif
  endif
endif
```
After discussions with JEDI developers it became apparent that the above approach is no longer feasible given JEDI design decisions.   Alternative methods must be developed to compute the analysis `delp` and hydrostatic `delz` increments.

The easiest way to achieve this end at present is to modify `ush/jediinc2fv3.py`.   This script currently runs after `fv3jedi_var.x` to create an analysis increment file using the variable names expected by the FV3GFS forecast model.   This issue documents the addition of these calculations.

**Comment by RussTreadon-NOAA** [comment]
Create GDASApp branch [`feature/delp_delz`](https://github.com/NOAA-EMC/GDASApp/tree/feature/delp_delz) as a copy of `develop` at [8e8d725](https://github.com/NOAA-EMC/GDASApp/commit/8e8d7257d6905108dfae820caa3ae1bf993e9269) for this issue.

**Comment by RussTreadon-NOAA** [comment]
`test_gdasapp_jedi_increment_to_fv3` fails due to changes in `jediinc2fv3.py`.    

```
(gdasapp) Orion-login-4[73] rtreadon$ ctest -R test_gdasapp_jedi_increment_to_fv3 --output-on-failure
Test project /work/noaa/da/Russ.Treadon/git/GDASApp/delp_delz/build
    Start 1574: test_gdasapp_jedi_increment_to_fv3
1/1 Test #1574: test_gdasapp_jedi_increment_to_fv3 ...***Failed    0.23 sec
usage: jediinc2fv3.py [-h] FV3background FV3JEDIincrement FV3increment
jediinc2fv3.py: error: the following arguments are required: FV3increment


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.80 sec

The following tests FAILED:
        1574 - test_gdasapp_jedi_increment_to_fv3 (Failed)
Errors while running CTest
```


The revised script now requires three arguments
1. `FV3background` - fv3 background file _(input)_
2. `FV3JEDIincrement` - jedi analysis increment file _(input)_
3. `FV3increment` - fv3 analysis increment file _(output)_

Previously only `FV3JEDIincrement` and  `FV3increment` were required.    `FV3background` is now required for calculation of the `delp` and hydrostatic `delz` increments.

The `FV3increment` file used by the ctest, `atminc_compress.nc4`, resides in `/home/ftp/emc/static_files/public/GDASApp/testdata`.   A corresponding background file needs to be added.  Alternatively a new pair of background and analysis increment files need to be added.

Follow up with @CoryMartin-NOAA as to how best to proceed.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA I have a tar file on EMC's FTP site that I keep updated with sample test data. We will need to create low resolution files for the unit tests. Once they are ready, I can upload them and create the new tar file.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , I will work on creating low resolution files.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , how did you create `atminc_compress.nc4`?  This is the increment file read by `test_gdasapp_jedi_increment_to_fv3`.    Knowing how `atminc` was created is useful information as I consider various ways to create a low resolution `atmges` and `atminc`.  We have C96L127 files but these are too large to store on EMC's FTP site and use for ctests.

**Comment by RussTreadon-NOAA** [comment]
As a test, run `jediinc2fv3.py` using output from a C96L127 `fv3jedi_var.x` run for 2021122106.   The gaussian grid file increment file generated by `fv3jedi_var.x` along with the previous cycle atmf006.nc background file were passed to `jediinc2fv3.py` via command line arguments.   

Shown below is a scatter plot of the delz_increment computed by a modified version of `fv3jedi_var.x` (_x-axis_) versus the delz_increment computed by `jediinc2fv3.py` (_y-axis_).  The scatter plot includes all points on the three dimension increment grids.
![delz_inc_scatter](https://user-images.githubusercontent.com/26926959/192047646-07b788ea-b80b-4289-b7e4-c35e720b60ef.png)

Similarly, here is a scatter plot for the delp_increments.
![delp_inc_scatter](https://user-images.githubusercontent.com/26926959/192048251-61039aeb-d7a0-45b3-bf28-0c66c37c9a22.png)

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA I'm pretty sure I used the FV3-JEDI latlon increment write capability to create this file from C12 test background RESTART files from FV3-JEDI. Then I think I ran it through one of the nco tools to use the maximum compression (deflate level 9). Do you plan to use cube sphere history files (still in development) or FMS RESTART files for the backgrounds/analysis? If the latter, we can probably 'borrow' from `fv3-jedi-data`

**Comment by RussTreadon-NOAA** [comment]
Low resolution gaussian grid background and increment files have been created by running `fv3jedi_test_tier1_letkf`.  The yaml file used in this ctest, `letkf.yaml`, was edited to write out the prior ensemble mean background and the posterior ensemble mean increment.   A python script was written to refactor the prior ensemble mean background file to better mimic a FV3GFS gaussian grid background file (e.g., include ak, bk, use FV3GFS atmfXXX.nc variable names, etc).  Both the refactored background and increment files were passed through `nccopy -d9` to create compressed counterparts:  `atmges_compress.nc4` and `atminc_compress.nc4`.    

`test_gdasapp_jedi_increment_to_fv3` was executed with the new atmges and atminc.   Script `jediinc2fv3.py`  at [c1e2a44](https://github.com/NOAA-EMC/GDASApp/commit/c1e2a44efda6206fc9de3069c5696342a687157f) successfully ran to completion. 
```
(gdasapp) Orion-login-1[48] rtreadon$ ctest -VV -R test_gdasapp_jedi_increment_to_fv3
UpdateCTestConfiguration  from :/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/DartConfiguration.tcl
Parse Config file:/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/DartConfiguration.tcl
Parse Config file:/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/DartConfiguration.tcl
Test project /work/noaa/da/Russ.Treadon/git/GDASApp/develop/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 2201
    Start 2201: test_gdasapp_jedi_increment_to_fv3

2201: Test command: /work2/noaa/da/python/opt/core/miniconda3/4.6.14/envs/gdasapp/bin/python3.7 "/work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py" "/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/test/testdata/atmges_compress.nc4" "/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/test/testdata/atminc_compress.nc4" "/work/noaa/da/Russ.Treadon/git/GDASApp/develop/build/test/testoutput/fv_increment.nc"
2201: Test timeout computed to be: 1500
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:88: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   ncout[vardict[name]][:] = ncin[name][:]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:84: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   ncout[vardict[name]][:] = ncin[name][0, ...]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:95: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   ps_ges = ncges.variables['pressfc'][:]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:96: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   t_ges = ncges.variables['tmp'][:]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:97: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   q_ges = ncges.variables['spfh'][:]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:99: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   ps_inc = ncin.variables['ps'][:]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:100: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   t_inc = ncin.variables[tinc][:]
2201: /work/noaa/da/Russ.Treadon/git/GDASApp/develop/ush/jediinc2fv3.py:101: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
2201: Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
2201:   q_inc = ncin.variables['sphum'][:]
1/1 Test #2201: test_gdasapp_jedi_increment_to_fv3 ...   Passed    0.48 sec

The following tests passed:
        test_gdasapp_jedi_increment_to_fv3

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   1.17 sec
(gdasapp) Orion-login-1[49] rtreadon$
```

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , two new files have been created for `test_gdasapp_jedi_increment_to_fv3`
- `atmges_compress.nc4` - fv3gfs like background file
- `atminc_compress.nc4` - fv3jedi analysis increment

These files are on Orion in `/work/noaa/stmp/rtreadon/ctest`.    They are also available on Hera in `/scratch2/NCEPDEV/stmp1/Russ.Treadon/ctest`.   We need to upload these files to emcrzdm directory `/home/ftp/emc/static_files/public/GDASApp/testdata`.  

I guess a new tarball also needs to be created.   Once this is done will `/GDASApp/develop/test/CMakeLists.txt` need the hash to be updated?

**Comment by RussTreadon-NOAA** [comment]
As a test, do the following:

- copy `atmges_compress.nc4` and `atminc_compress.nc4` to emcrzdm `/home/ftp/emc/static_files/public/GDASApp_test/testdata`
- create compressed tarball `gdasapp-fix-232c89.tgz` in emcrzdm `/home/ftp/emc/static_files/public/GDASApp_test`
- cd into a working copy of `feature/delp_delz` and replace `static_files/public/GDASApp` with `static_files/public/GDASApp_test` in `test/CMakeLists.txt`
- build GDASApp
- execute ctest `test_gdasapp_jedi_increment_to_fv3`.   Test successfully ran to completion.   Check ctest output `fv3_increment.nc` to confirm that `delp_inc` and `delz_inc` present and contain non-zero values.

We we still need to get the new `atmges_compress.nc4` and `atminc_compress.nc4` into the official GDASApp ctest tarball, the success of this exercise is sufficient to warrant opening a daft PR to request merger of `feature/delp_delz` into `develop`.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA yep that is the process. As long as that is the sha256 sum, that is 'official' enough for me. Once the tarball is on RZDM, you change the hash in `test/CMakeLists.txt` to the full sha256 hash and it will download that file as part of the cmake/build process.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA try using the hash `523cd3b8a1dbb30adc22f0302a0588775f9ac44b8e3463273abf4f1e63304d81` and see if the new tarball has what you need in it

**Comment by RussTreadon-NOAA** [comment]
Thank you, @CoryMartin-NOAA , for updating the GDASApp tarball.   

Unfortunately, `gdasapp-fix-523cd3.tgz` does not work.   GDASApp `parm/atm/variational/3dvar_dripcg.yaml` lists `ps` as an analysis variable.   To mimic `fv3jedi_var.x` output we should have surface pressure in the fv3jedi increment file.   The `atminc_compress.nc4` in `gdasapp-fix-523cd3.tgz` contains `DELP`.   

I created a new `atminc_compress.nc4` with `ps`.   Hera directory /scratch2/NCEPDEV/stmp1/Russ.Treadon/ctest contains the new `atminc_compress.nc4`.   Can we replace `atminc_compress.nc4` in emcrzdm `atminc_compress.nc4` with the new file and create a new gdasapp-fix tarball?

**Comment by CoryMartin-NOAA** [comment]
For some reason, I am really struggling with getting either rzdm to talk to Hera or vice versa for SCP. Any chance the file you had in GDASApp_test on RZDM is the right atminc? If so, try this hash: `2f85f9469ee0788b3765a667fa6e9cd52b599dfb986b7dc773aaf4bcdaf05b4b`

**Comment by RussTreadon-NOAA** [comment]
Thank you, @CoryMartin-NOAA !   The new tarball, `gdasapp-fix-2f85f9.tgz` works.   `test_gdasapp_jedi_increment_to_fv3` passed.

**Comment by RussTreadon-NOAA** [comment]
PR #139 was merged into `develop` at [1ef214b](https://github.com/NOAA-EMC/GDASApp/commit/1ef214b85f6904967b62e6ebc21c0c7815872f86).  Close this issue as **complete**.
